autosklearn/automl.py

```diff
@@ -1834,15 +1834,52 @@ def get_models_with_weights(self):
         return self.ensemble_.get_models_with_weights(self.models_)
 
     def show_models(self):
-        models_with_weights = self.get_models_with_weights()
-
-        with io.StringIO() as sio:
-            sio.write("[")
-            for weight, model in models_with_weights:
-                sio.write("(%f, %s),\n" % (weight, model))
-            sio.write("]")
-
-            return sio.getvalue()
+        """ Returns a dictionary containing models and their information
+            where model_id/run_id is the key """
+
+        ensemble_dict={} 
+        
+        def has_key(rv, key):
+            return rv.additional_info and key in rv.additional_info
+        
+        table_dict={}
+        for rkey, rval in self.runhistory_.data.items():
+            if has_key(rval, 'num_run'):
+                table_dict[rval.additional_info['num_run']]={
+                        'model_id': rval.additional_info['num_run'],
+                        'cost': rval.cost}
+                
+        for i, weight in enumerate(self.ensemble_.weights_):
+            (_, model_id, _) = self.ensemble_.identifiers_[i]
+            table_dict[model_id]['ensemble_weight']=weight
+        
+        table=pd.DataFrame.from_dict(table_dict, orient='index')
+        table.sort_values(by='cost', inplace=True)
+        table['rank']=range(1,len(table)+1)
+
+        for (_, model_id, _), model in self.models_.items():
+            model_dict={} #Empty model dictionary
+                
+            #Inserting rank and ensemble weight
+            model_dict['model_id']=table.loc[model_id]['model_id']
+            model_dict['rank']=table.loc[model_id]['rank']
+            model_dict['ensemble_weight']=table.loc[model_id]['ensemble_weight']
+            
+            # The steps in the models pipeline will be saved in the dictionary as follows: 
+            # 'data_preprocessing': DataPreprocessor,
+            # 'balancing': Balancing,
+            # 'feature_preprocessor': FeaturePreprocessorChoice,
+            # 'classifier': ClassifierChoice -> autosklearn wrapped model
+            for step in model.steps:
+                model_dict[step[0]]=step[1]
```

Comment:
I would spell out the nested indexing part, i.e. `model_dict[step[0]]` into two steps to be more clear.

```python
step_name, step_object = step
model_dict[step_name] = step_object
```

This is probably a bit more confusing if you're unaware of the `dict` constructor being able to take lists of tuples but it's also possible to do the following. With the comment above it should be okay though.
```python
steps = dict(model.steps)
model_dict.update(steps)

# Example
mysteps = [ ('a', 1), ('b', 2), ..., ('z', 26) ]
mysteps_as_dict = dict(mysteps)
print(mysteps_as_dict)
# {
#     'a': 1,
#     'b': 2,
#     ...
#     'z': 26
# }
```

As a bonus, here's a little handy trick to convert two seperate lists to a dictionary.
```python
letters = ['a', 'b', 'c']
numbers = [1, 2, 3]
d = dict(zip(letters, numbers)
# {
#    'a': 1,
#    'b': 2,
#    'c': 3
# }
```

Reply:
This is a solid tip, thanks for sharing!